[stdlib] [SR-4005] Allow heterogenous comparisons in elementsEqual

Cbieniak opened the pull request:
When supplying a predicate to compare, type equivalence of the two sequences are not necessary

Resolves [SR-4005](https://bugs.swift.org/browse/SR-4005).

Cbieniak committed:
[SR-4005] Allow heterogenous comparisons in elementsEqual

When a user is supplying a predicate to compare the type equivalence
isn’t required

Cbieniak commented:
@swift-ci Please smoke test

jrose-apple commented on stdlib/public/core/SequenceAlgorithms.swift.gyb:
```diff
@@ -353,12 +353,14 @@ ${equivalenceExplanation}
-    OtherSequence: Sequence,
+    OtherSequence: Sequence${" {" if preds else ","}
+%   if not preds:
     OtherSequence.${GElement} == ${GElement} {
```
@nkcsgexi It is concerning that the change in the generic requirements was not caught by swift-api-digester.

nkcsgexi replied:
Right, the diff tool is not comparing generic requirements now.

airspeedswift commented:
@swift-ci Please test

airspeedswift reviewed (changes requested):
Could you add some changes to test some comparisons of different element types using a closure? (for example, comparing `["0","1","2"]` to `0..<4`)

Cbieniak committed:
elementsEqualWithPredicate tests

Compares a string of a number with an integer value by using the
elementsEqualPredicate closure

Cbieniak commented:
@airspeedswift I've written a couple comparing the string version of ints being converted. Is this appropriate?

Cbieniak committed:
Update test expectations to use new sequence element types

Cbieniak commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

Cbieniak commented:
@jrose-apple thanks for the review. The only error I could see in the linux build was 

```TestFoundation/TestNSURLSession.swift:245: error: TestURLSession.test_downloadTaskWithURLAndHandler : XCTAssertEqual failed: ("cannotConnectToHost") is not equal to ("timedOut") - Unexpected error code```

my best guess would be that its an issue with the linux instance rather than the pr, Should we just rerun the tests or did you have advice on how id go about fixing this?

jrose-apple commented:
Ah, yeah, seems unlikely. Let's try again.

(This still isn't a review, though. @airspeedswift needs to accept the new version.)

@swift-ci Please test Linux platform

airspeedswift commented:
@swift-ci please test and merge

jrose-apple commented:
@swift-ci Please test source compatibility

airspeedswift commented:
Oops, thanks @jrose. Tho it'd be pretty weird if this caused anything to fail.

airspeedswift commented:
@swift-ci please clean test linux platform

jrose-apple commented:
The compat suite failures are in AsyncNinja, which we think is due to Pavel's recent change. Not related to this commit.

airspeedswift commented:
Hi @Cbieniak – sorry, looks like I just caused a conflict with another commit. Could you rebase and hopefully we can land this? Thanks!

Cbieniak commented:
@airspeedswift No worries at all mate.

airspeedswift commented:
@swift-ci please test

airspeedswift commented:
@swift-ci please smoke test compiler performance

airspeedswift commented:
@swift-ci please test source compatibility

airspeedswift commented:
Ah. Looks like one of the hardcoded tests that check completion needs tweaking.

airspeedswift commented:
@swift-ci please test

Cbieniak commented:
Looks like an issue with the complete_from_stdlib.swift. Ill fix it when I get a chance

airspeedswift commented:
Thanks @Cbieniak!

Cbieniak committed:
Update hardcoded test to reference sequence

Cbieniak commented:
Should be all sorted now

airspeedswift commented:
@swift-ci please test

airspeedswift commented:
@swift-ci please test macOS platform
